### ISSUE TITLE
[MNG-5437] Do not load plugins from snapshot JARs

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6497]
**[Jesse N. Glick](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jglick)** opened **[MNG-5437](https://issues.apache.org/jira/browse/MNG-5437?redirect=false)** and commented

See [mailing list discussion](http://mail-archives.apache.org/mod_mbox/www-builds/201302.mbox/browser) for background. When a reactor build creates a plugin JAR and then loads classes from it, a JVM crash is a possibility:

```none
[INFO] Installing /export/home/hudson/hudson-slave/workspace/sis-jdk7/sis-build-helper/target/sis-build-helper-0.3-jdk7-SNAPSHOT.jar to /export/home/hudson/hudson-slave/maven-repositories/0/org/apache/sis/sis-build-helper/0.3-jdk7-SNAPSHOT/sis-build-helper-0.3-jdk7-SNAPSHOT.jar
...moving on to another project in the reactor...
[INFO] --- sis-build-helper:0.3-jdk7-SNAPSHOT:collect-jars (default) @ sis-utility ---
#
1. A fatal error has been detected by the Java Runtime Environment:
```

This could happen for example if two people were running the same build concurrently; if one Maven process is trying to load classes from the JAR at the same time as another is rebuilding that JAR, there will be a problem. On Windows, one of the processes will simply fail with an I/O error due to mandatory file locks. On Unix, the OS will not stop the clash in this way, but there will be a worse issue: `libzip.so` will be trying to load a ZIP entry based on a stale copy of the ZIP index, so it loads bogus data, and (being not very robustly written) crashes, taking the whole JVM with it.

Since this class of problem is most likely to occur when loading classes from `SNAPSHOT` JARs in the local repository, I would suggest Maven simply not do this. Instead, use `File.createTempFile` to make a new `/tmp/*.jar`; use `deleteOnExit` to clean it up at process end (or if possible delete it sooner, e.g. upon conclusion of the reactor build, or at any time when it can be determined that mojo execution from this JAR must have ceased); copy the snapshot JAR from the local repo to this temp location; and load from there.

This means that problems will be limited to the much less likely event that someone is rebuilding the JAR during the very short period during which the file copy happens. And on Unix, the result might be a corrupted JAR file (and thus class loading errors which fail the build somehow), but not a mysterious JVM crash.

If this proposal sounds good, let me know and I can try to write a patch. (GitHub pull request would be easiest; I am an Apache committer.)

---

**Affects:** 3.0.4

**Remote Links:**
- [Bug Report with Oracle: JDK-4425695
](https://bugs.java.com/view_bug.do?bug_id=4425695)
- [Stackoveflow Reference #1
](https://stackoverflow.com/q/1313071/696632)
- [Stackoveflow Reference #2
](https://stackoverflow.com/q/1313071/696632)
- [Implementation Ticket w/ Oracle #1
](https://bugs.openjdk.java.net/browse/JDK-8142508)
- [Implementation Ticket w/ Oracle #2
](https://bugs.openjdk.java.net/browse/JDK-8145260)

0 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Martin Desruisseaux](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=desruisseaux)** commented

The proposed patch may help in making Maven build more robust when running builds in parallel, thanks a lot for proposing it. But I wonder if this is a common cases? In the particular case of the JVM crash reported in the SIS project, there is no parallel build, and I don't see which process could be cobbling the JAR file at the time of the JVM crash... So it is quite possible that the JVM crash has another cause. One think we could do first would be to upgrade JDK7 on the Jenkins server (which is 18 months old), then see if the crashes are still there.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I have experienced the very same issue with Java 8 and recent Maven 3.5.x on FreeBSD in  Maven Integration Testing Suite. One would actually need to rework ITs for this.

Give that one can disable `mmap` via system property and Java 9 does not use native code anymore for this. Closing as won't fix.
